mqtt/mqtt-broker/src/persist.rs

```diff
@@ -136,6 +145,211 @@ impl FileFormat for BincodeFormat {
     }
 }
 
+#[derive(Clone, Debug, Default)]
+pub struct ConsolidatedStateFormat;
+
+impl ConsolidatedStateFormat {
+    pub fn new() -> Self {
+        Self
+    }
+}
+
+impl FileFormat for ConsolidatedStateFormat {
+    type Error = Error;
+
+    fn load<R: Read>(&self, reader: R) -> Result<BrokerState, Self::Error> {
+        let decoder = GzDecoder::new(reader);
+        fail_point!("bincodeformat.load.deserialize_from", |_| {
+            Err(Error::from(ErrorKind::Persist(ErrorReason::Deserialize)))
+        });
+        let state = bincode::deserialize_from(decoder)
+            .context(ErrorKind::Persist(ErrorReason::Deserialize))?;
+
+        let state = Resolver::resolve_state(state);
+        Ok(state)
+    }
+
+    fn store<W: Write>(&self, writer: W, state: BrokerState) -> Result<(), Self::Error> {
+        let state = Consolidator::consolidate_state(state);
+
+        let encoder = GzEncoder::new(writer, Compression::default());
+        fail_point!("bincodeformat.store.serialize_into", |_| {
+            Err(Error::from(ErrorKind::Persist(ErrorReason::Deserialize)))
+        });
+        bincode::serialize_into(encoder, &state)
+            .context(ErrorKind::Persist(ErrorReason::Serialize))?;
+        Ok(())
+    }
+}
+
+struct Consolidator {
+    payloads: HashMap<Bytes, u64>,
+    curr_id: u64,
+}
+
+impl Consolidator {
```

Comment:
it's hard to get quick understanding what is going on here. It would be great if we can add code docs that explain what is Consolidator and what problem it tries to solve.

Reply:
denis's recommended changed should make it better